testsuite failure (LinkNotFoundError)

Hello, since some days Debian and Ubuntu testsuite have started failing... Probably httpbin changed some bits?
see: https://ci.debian.net/packages/p/python-mechanicalsoup/

and the logs: https://ci.debian.net/packages/p/python-mechanicalsoup/unstable/amd64/

```
Testing with python3.6:
============================= test session starts ==============================
platform linux -- Python 3.6.6rc1+, pytest-3.3.2, py-1.5.3, pluggy-0.6.0 -- /usr/bin/python3.6
cachedir: .cache
rootdir: /tmp/autopkgtest-lxc.vt1vanff/downtmp/autopkgtest_tmp, inifile:
plugins: mock-1.7.1
collecting ... collected 64 items

tests/test_browser.py::test_submit_online PASSED                         [  1%]
tests/test_browser.py::test__request PASSED                              [  3%]
tests/test_browser.py::test__request_file PASSED                         [  4%]
tests/test_browser.py::test_no_404 PASSED                                [  6%]
tests/test_browser.py::test_404 PASSED                                   [  7%]
tests/test_browser.py::test_set_cookiejar PASSED                         [  9%]
tests/test_browser.py::test_get_cookiejar PASSED                         [ 10%]
tests/test_browser.py::test_post PASSED                                  [ 12%]
tests/test_form.py::test_submit_online PASSED                            [ 14%]
tests/test_form.py::test_submit_set PASSED                               [ 15%]
tests/test_form.py::test_choose_submit[preview] PASSED                   [ 17%]
tests/test_form.py::test_choose_submit[save] PASSED                      [ 18%]
tests/test_form.py::test_choose_submit[cancel] PASSED                    [ 20%]
tests/test_form.py::test_choose_submit_fail[not found] PASSED            [ 21%]
tests/test_form.py::test_choose_submit_fail[found] PASSED                [ 23%]
tests/test_form.py::test_choose_submit_multiple_match PASSED             [ 25%]
tests/test_form.py::test_form_noaction PASSED                            [ 26%]
tests/test_form.py::test_form_action PASSED                              [ 28%]
tests/test_form.py::test_set_select[default] PASSED                      [ 29%]
tests/test_form.py::test_set_select[selected] PASSED                     [ 31%]
tests/test_form.py::test_set_select_multiple[select one (str)] PASSED    [ 32%]
tests/test_form.py::test_set_select_multiple[select one (tuple)] PASSED  [ 34%]
tests/test_form.py::test_set_select_multiple[select two] PASSED          [ 35%]
tests/test_form.py::test_form_not_found PASSED                           [ 37%]
tests/test_form.py::test_form_check_uncheck PASSED                       [ 39%]
tests/test_form.py::test_form_print_summary PASSED                       [ 40%]
tests/test_form.py::test_issue180 PASSED                                 [ 42%]
tests/test_stateful_browser.py::test_request_forward PASSED              [ 43%]
tests/test_stateful_browser.py::test_submit_online PASSED                [ 45%]
tests/test_stateful_browser.py::test_no_404 PASSED                       [ 46%]
tests/test_stateful_browser.py::test_404 PASSED                          [ 48%]
tests/test_stateful_browser.py::test_user_agent PASSED                   [ 50%]
tests/test_stateful_browser.py::test_open_relative PASSED                [ 51%]
tests/test_stateful_browser.py::test_links PASSED                        [ 53%]
tests/test_stateful_browser.py::test_submit_btnName[input] PASSED        [ 54%]
tests/test_stateful_browser.py::test_submit_btnName[button] PASSED       [ 56%]
tests/test_stateful_browser.py::test_get_set_debug PASSED                [ 57%]
tests/test_stateful_browser.py::test_list_links PASSED                   [ 59%]
tests/test_stateful_browser.py::test_launch_browser PASSED               [ 60%]
tests/test_stateful_browser.py::test_find_link PASSED                    [ 62%]
tests/test_stateful_browser.py::test_verbose PASSED                      [ 64%]
tests/test_stateful_browser.py::test_new_control PASSED                  [ 65%]
tests/test_stateful_browser.py::test_form_noaction PASSED                [ 67%]
tests/test_stateful_browser.py::test_form_noname PASSED                  [ 68%]
tests/test_stateful_browser.py::test_form_multiple PASSED                [ 70%]
tests/test_stateful_browser.py::test_upload_file PASSED                  [ 71%]
tests/test_stateful_browser.py::test_with PASSED                         [ 73%]
tests/test_stateful_browser.py::test_select_form_nr PASSED               [ 75%]
tests/test_stateful_browser.py::test_select_form_tag_object PASSED       [ 76%]
tests/test_stateful_browser.py::test_referer_follow_link FAILED          [ 78%]
tests/test_stateful_browser.py::test_referer_submit PASSED               [ 79%]
tests/test_stateful_browser.py::test_referer_submit_headers PASSED       [ 81%]
tests/test_stateful_browser.py::test_link_arg_text PASSED                [ 82%]
tests/test_stateful_browser.py::test_link_arg_regex PASSED               [ 84%]
tests/test_stateful_browser.py::test_link_arg_multiregex PASSED          [ 85%]
tests/test_stateful_browser.py::test_download_link FAILED                [ 87%]
tests/test_stateful_browser.py::test_download_link_nofile FAILED         [ 89%]
tests/test_stateful_browser.py::test_download_link_to_existing_file FAILED [ 90%]
tests/test_stateful_browser.py::test_download_link_404 PASSED            [ 92%]
tests/test_stateful_browser.py::test_download_link_referer PASSED        [ 93%]
tests/test_stateful_browser.py::test_refresh_open PASSED                 [ 95%]
tests/test_stateful_browser.py::test_refresh_follow_link PASSED          [ 96%]
tests/test_stateful_browser.py::test_refresh_form_not_retained PASSED    [ 98%]
tests/test_stateful_browser.py::test_refresh_error PASSED                [100%]

=================================== FAILURES ===================================
___________________________ test_referer_follow_link ___________________________

httpbin = <utils.HttpbinRemote object at 0x7ff987009630>

    def test_referer_follow_link(httpbin):
        browser = mechanicalsoup.StatefulBrowser()
        browser.open(httpbin.url)
>       response = browser.follow_link("/headers")

tests/test_stateful_browser.py:371: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:324: in follow_link
    link = self._find_link_internal(link, args, kwargs)
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:300: in _find_link_internal
    return self.find_link(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mechanicalsoup.stateful_browser.StatefulBrowser object at 0x7ff9870096a0>
args = (), kwargs = {'url_regex': '/headers'}, links = []

    def find_link(self, *args, **kwargs):
        """Find and return a link, as a bs4.element.Tag object.
    
            The search can be refined by specifying any argument that is accepted
            by :func:`links`. If several links match, return the first one found.
    
            If no link is found, raise :class:`LinkNotFoundError`.
            """
        links = self.links(*args, **kwargs)
        if len(links) == 0:
>           raise LinkNotFoundError()
E           mechanicalsoup.utils.LinkNotFoundError

/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:274: LinkNotFoundError
------------------------------ Captured log call -------------------------------
connectionpool.py          208 DEBUG    Starting new HTTP connection (1): httpbin.org
connectionpool.py          396 DEBUG    http://httpbin.org:80 "GET / HTTP/1.1" 200 8344
______________________________ test_download_link ______________________________

httpbin = <utils.HttpbinRemote object at 0x7ff98701c240>

    def test_download_link(httpbin):
        """Test downloading the contents of a link to file."""
        browser = mechanicalsoup.StatefulBrowser()
        browser.open(httpbin.url)
        tmpdir = tempfile.mkdtemp()
        tmpfile = tmpdir + '/nosuchfile.png'
        current_url = browser.get_url()
        current_page = browser.get_current_page()
>       response = browser.download_link(file=tmpfile, link='image/png')

tests/test_stateful_browser.py:453: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:347: in download_link
    link = self._find_link_internal(link, args, kwargs)
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:300: in _find_link_internal
    return self.find_link(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mechanicalsoup.stateful_browser.StatefulBrowser object at 0x7ff98701cd68>
args = (), kwargs = {'url_regex': 'image/png'}, links = []

    def find_link(self, *args, **kwargs):
        """Find and return a link, as a bs4.element.Tag object.
    
            The search can be refined by specifying any argument that is accepted
            by :func:`links`. If several links match, return the first one found.
    
            If no link is found, raise :class:`LinkNotFoundError`.
            """
        links = self.links(*args, **kwargs)
        if len(links) == 0:
>           raise LinkNotFoundError()
E           mechanicalsoup.utils.LinkNotFoundError

/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:274: LinkNotFoundError
------------------------------ Captured log call -------------------------------
connectionpool.py          208 DEBUG    Starting new HTTP connection (1): httpbin.org
connectionpool.py          396 DEBUG    http://httpbin.org:80 "GET / HTTP/1.1" 200 8344
__________________________ test_download_link_nofile ___________________________

httpbin = <utils.HttpbinRemote object at 0x7ff9870646d8>

    def test_download_link_nofile(httpbin):
        """Test downloading the contents of a link without saving it."""
        browser = mechanicalsoup.StatefulBrowser()
        browser.open(httpbin.url)
        current_url = browser.get_url()
        current_page = browser.get_current_page()
>       response = browser.download_link(link='image/png')

tests/test_stateful_browser.py:472: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:347: in download_link
    link = self._find_link_internal(link, args, kwargs)
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:300: in _find_link_internal
    return self.find_link(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mechanicalsoup.stateful_browser.StatefulBrowser object at 0x7ff987064710>
args = (), kwargs = {'url_regex': 'image/png'}, links = []

    def find_link(self, *args, **kwargs):
        """Find and return a link, as a bs4.element.Tag object.
    
            The search can be refined by specifying any argument that is accepted
            by :func:`links`. If several links match, return the first one found.
    
            If no link is found, raise :class:`LinkNotFoundError`.
            """
        links = self.links(*args, **kwargs)
        if len(links) == 0:
>           raise LinkNotFoundError()
E           mechanicalsoup.utils.LinkNotFoundError

/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:274: LinkNotFoundError
------------------------------ Captured log call -------------------------------
connectionpool.py          208 DEBUG    Starting new HTTP connection (1): httpbin.org
connectionpool.py          396 DEBUG    http://httpbin.org:80 "GET / HTTP/1.1" 200 8344
_____________________ test_download_link_to_existing_file ______________________

httpbin = <utils.HttpbinRemote object at 0x7ff987076d68>

    def test_download_link_to_existing_file(httpbin):
        """Test downloading the contents of a link to an existing file."""
        browser = mechanicalsoup.StatefulBrowser()
        browser.open(httpbin.url)
        tmpdir = tempfile.mkdtemp()
        tmpfile = tmpdir + '/existing.png'
        with open(tmpfile, "w") as f:
            f.write("initial content")
        current_url = browser.get_url()
        current_page = browser.get_current_page()
>       response = browser.download_link('image/png', tmpfile)

tests/test_stateful_browser.py:492: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:347: in download_link
    link = self._find_link_internal(link, args, kwargs)
/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:300: in _find_link_internal
    return self.find_link(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mechanicalsoup.stateful_browser.StatefulBrowser object at 0x7ff987076dd8>
args = (), kwargs = {'url_regex': 'image/png'}, links = []

    def find_link(self, *args, **kwargs):
        """Find and return a link, as a bs4.element.Tag object.
    
            The search can be refined by specifying any argument that is accepted
            by :func:`links`. If several links match, return the first one found.
    
            If no link is found, raise :class:`LinkNotFoundError`.
            """
        links = self.links(*args, **kwargs)
        if len(links) == 0:
>           raise LinkNotFoundError()
E           mechanicalsoup.utils.LinkNotFoundError

/usr/lib/python3/dist-packages/mechanicalsoup/stateful_browser.py:274: LinkNotFoundError
------------------------------ Captured log call -------------------------------
connectionpool.py          208 DEBUG    Starting new HTTP connection (1): httpbin.org
connectionpool.py          396 DEBUG    http://httpbin.org:80 "GET / HTTP/1.1" 200 8344
===================== 4 failed, 60 passed in 6.69 seconds ======================
```
Do you have any update?

--- a/tests/test_stateful_browser.py
+++ b/tests/test_stateful_browser.py
@@ -6,7 +6,12 @@ import mechanicalsoup
 import sys
 import re
 from bs4 import BeautifulSoup
-from utils import setup_mock_browser, prepare_mock_browser, mock_get
+from utils import (
+    setup_mock_browser,
+    prepare_mock_browser,
+    mock_get,
+    open_legacy_httpbin
+)
 import pytest
 import webbrowser
 
@@ -367,11 +372,12 @@ def test_select_form_tag_object():
 
 def test_referer_follow_link(httpbin):
     browser = mechanicalsoup.StatefulBrowser()
-    browser.open(httpbin.url)
+    open_legacy_httpbin(browser, httpbin)
+    start_url = browser.get_url()
     response = browser.follow_link("/headers")
     referer = response.json()["headers"]["Referer"]
     actual_ref = re.sub('/*$', '', referer)
-    expected_ref = re.sub('/*$', '', httpbin.url)
+    expected_ref = re.sub('/*$', '', start_url)
     assert actual_ref == expected_ref
 
 
@@ -445,7 +451,7 @@ def file_get_contents(filename):
 def test_download_link(httpbin):
     """Test downloading the contents of a link to file."""
     browser = mechanicalsoup.StatefulBrowser()
-    browser.open(httpbin.url)
+    open_legacy_httpbin(browser, httpbin)
     tmpdir = tempfile.mkdtemp()
     tmpfile = tmpdir + '/nosuchfile.png'
     current_url = browser.get_url()
@@ -466,7 +472,7 @@ def test_download_link(httpbin):
 def test_download_link_nofile(httpbin):
     """Test downloading the contents of a link without saving it."""
     browser = mechanicalsoup.StatefulBrowser()
-    browser.open(httpbin.url)
+    open_legacy_httpbin(browser, httpbin)
     current_url = browser.get_url()
     current_page = browser.get_current_page()
     response = browser.download_link(link='image/png')
@@ -482,7 +488,7 @@ def test_download_link_nofile(httpbin):
 def test_download_link_to_existing_file(httpbin):
     """Test downloading the contents of a link to an existing file."""
     browser = mechanicalsoup.StatefulBrowser()
-    browser.open(httpbin.url)
+    open_legacy_httpbin(browser, httpbin)
     tmpdir = tempfile.mkdtemp()
     tmpfile = tmpdir + '/existing.png'
     with open(tmpfile, "w") as f:

--- a/tests/utils.py
+++ b/tests/utils.py
@@ -76,3 +76,20 @@ class HttpbinRemote:
 
     def __add__(self, x):
         return self.url + x
+
+
+def open_legacy_httpbin(browser, httpbin):
+    """Opens the start page of httpbin (given as a fixture). Tries the
+    legacy page (available only on recent versions of httpbin), and if
+    it fails fall back to the main page (which is JavaScript-only in
+    recent versions of httpbin hence usable for us only on old
+    versions).
+    """
+    try:
+        response = browser.open(httpbin + "/legacy")
+        if response.status_code == 404:
+            # The line above may or may not have raised the expection
+            # depending on raise_on_404. Raise it unconditionally now.
+            raise mechanicalsoup.LinkNotFoundError()
+    except mechanicalsoup.LinkNotFoundError:
+        browser.open(httpbin.url)

